lychee-lib/src/filter/mod.rs

```diff
@@ -1,72 +1,140 @@
-    pub(crate) scheme: Option<String>,
+    pub includes: Option<Includes>,
+    pub excludes: Option<Excludes>,
+    // TODO: accept multiple scheme
+    // TODO: includes scheme and excludes scheme
+    // TODO: excludes_mail should be merged to excludes scheme
+    // allowed scheme
+    pub scheme: Option<String>,
+    /// Example: 192.168.0.1
+    pub exclude_private_ips: bool,
+    /// Example: 169.254.0.0
+    pub exclude_link_local_ips: bool,
+    /// For IPv4: 127.0.0.1/8
+    /// For IPv6: ::1/128
+    pub exclude_loopback_ips: bool,
+    /// Example: octocat@github.com
+    pub exclude_mail: bool,
 }
 
 impl Filter {
-    pub fn new(
-        includes: Option<Includes>,
-        excludes: Option<Excludes>,
-        scheme: Option<String>,
-    ) -> Self {
-        Filter {
-            includes: includes.unwrap_or_default(),
-            excludes: excludes.unwrap_or_default(),
-            scheme,
+    #[inline]
+    pub fn is_mail_excluded(&self, uri: &Uri) -> bool {
+        uri.is_mail() && self.exclude_mail
+    }
+
+    pub fn is_ip_excluded(&self, uri: &Uri) -> bool {
+        match uri.host_ip() {
+            Some(ip_addr) if self.exclude_loopback_ips && ip_addr.is_loopback() => true,
+            // Note: in a pathological case, an IPv6 address can be IPv4-mapped
+            //       (IPv4 address embedded in a IPv6).  We purposefully
+            //       don't deal with it here, and assume if an address is IPv6,
+            //       we shouldn't attempt to map it to IPv4.
+            //       See: https://tools.ietf.org/html/rfc4291#section-2.5.5.2
+            Some(IpAddr::V4(v4_addr)) if self.exclude_private_ips && v4_addr.is_private() => true,
+            Some(IpAddr::V4(v4_addr)) if self.exclude_link_local_ips && v4_addr.is_link_local() => {
+                true
+            }
+            _ => false,
         }
     }
 
+    #[inline]
+    pub fn is_scheme_excluded(&self, uri: &Uri) -> bool {
+        matches!(self.scheme, Some(ref scheme) if scheme != uri.scheme())
+    }
+
+    #[inline]
+    fn is_includes_empty(&self) -> bool {
+        !matches!(self.includes, Some(ref includes) if !includes.is_empty())
+    }
+
+    #[inline]
+    fn is_excludes_empty(&self) -> bool {
+        !matches!(self.excludes, Some(ref excludes) if !excludes.is_empty())
+    }
+
+    #[inline]
+    fn is_includes_match(&self, input: &str) -> bool {
+        matches!(self.includes, Some(ref includes) if includes.is_match(input))
+    }
+
+    #[inline]
+    fn is_excludes_match(&self, input: &str) -> bool {
+        matches!(self.excludes, Some(ref excludes) if excludes.is_match(input))
+    }
+
+    /// Determine whether a given [`Uri`] should be excluded.
+    ///
+    /// # Details
+    ///
+    /// 1. If any of the following conditions is met, the URI is excluded:
+    ///   - If it's a mail address and it's configured to ignore mail addresses.
+    ///   - If the IP address belongs to a type that is configured to exclude.
```

Comment:
is configured to exclude -> is configured to be excluded

Reply:
No, the grammar is correct.

Consider `there's work to be done` and `there's work to do`.
Active infinitive focuses on the agent of action.
This is the documentation of the behaviour of a `Filter`. So the focus in on `Filter`.

If you use passive infinitive, the focus transferred to the receiver, which is the `Url`.